### PR TITLE
fix: improve MathML rendering in reports module

### LINKED ISSUE #191: Improve MathML rendering in report
Formulas are currently not rendered very nicely in the SBML reports.
This uses currently Conent MathML to Presention MathML conversion via the MathJax XLS stylesheets. Unfortunately this does not work very well.
I added helper functions which allow to directly convert `ASTNode` to `Latex` in the `mathml.py`.
The report should be updated so that ASTNode, i.e. math is rendered as latex. 
This requires updating the math information in the report and using the latex string with addition `$${latex_str}$$` which can be rendered by MathJax. See the `mathml.html` for an example.

### PR DESCRIPTION
**Pull request by sankha555**
* [x] fix #191 
* [x] improved MathML rendering in reports module by adding support for Latex and MathJax rendering
* [x] tests passed

### CODE REVIEW

**Comment by sankha555** [comment]
Added the preliminary version for the patch by adding the ```type``` parameter in the ```math()``` method and latex rendering support.  Adding the test fixes and ```math``` parameter to ```create_report``` now.

File: src/sbmlutils/report/sbmlreport.py [hunk 1]
```diff
@@ -38,6 +38,7 @@ def create_reports(
     template: str = "report.html",
     promote: bool = False,
     validate: bool = True,
+    math_type: str = "cmathml",
```
Comment (matthiaskoenig) [inline src/sbmlutils/report/sbmlreport.py:41]: You also must document the new parameter in the doc string.

**Comment by sankha555** [comment]
@matthiaskoenig I have made a PR for the issue. Please take a look at it. I have added the following things in two commits:

1. ```type``` parameter to ```math()``` in ```sbmlreport.py```. I have also added a pytest ```test_latex.py``` to specifically test for the reprissilator model using latex rendering. It works now! Attaching some of the Latex rendered formulae in the HTML report below:

image

image

2. In the second commit, I have added a ```math_type``` parameter to the ```create_reports```, ```create_report``` and ```create_html``` functions and all other ```listOf___ _ dicts``` functions from where the ```math()``` method is called. If you think this addition is suitable, can we add it to the customisation of the create_reports method? I have kept it as a separate commit for now, if it is needed to be removed, I would remove the commit. 

Also, there is a windows specific test that is failing on the CI/CD run. Could this be an error due to environment specific dependencies because the macOS and ubuntu runs have passed? The tests are also passing on my local machine (macOS Mojave 10.14).

 Please let me know your feedback and suggestions about the patch and please let me know if I should make any changes to it. Thanks!

File: src/sbmlutils/report/sbmlreport.py [hunk 2]
```diff
@@ -73,6 +75,7 @@ def create_report(
     output_dir: Path,
     promote: bool = False,
     template: str = "report.html",
+    math_type: str = "cmathml",
```
Comment (matthiaskoenig) [inline src/sbmlutils/report/sbmlreport.py:78]: document in doc string and add a sanity check for the input parameter along the line of
```
math_types = ["cmathml", "pmathml", "latex"]
if math_type not in math_types:
    raise ValueError(f"math_type '{math_type}' not in supported types: '{math_types}')
```

File: src/sbmlutils/report/sbmlreport.py [hunk 3]
```diff
@@ -225,18 +229,18 @@ def _create_html(
             "model": model_dict(model),
             "submodels": listOfSubmodels_dict(model),
             "ports": listOfPorts_dict(model),
-            "functions": listOfFunctions_dict(model),
+            "functions": listOfFunctions_dict(model, math_type),
```
Comment (matthiaskoenig) [inline src/sbmlutils/report/sbmlreport.py:232]: This is okay for now, but we have to refactor the report module into a class. We have to start writing the proposal and make the design of the new SBML report. I.e., there has to be a major refactoring in new components which have to work together.

File: src/sbmlutils/test/report/test_latex.py [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+"""
+import pytest
+
+from sbmlutils.report import sbmlreport
+from sbmlutils.test import REPRESSILATOR_SBML
+
+
+@pytest.mark.parametrize("source", [REPRESSILATOR_SBML])
+def test_report(source, tmp_path):
+    sbmlreport.create_report(sbml_path=source, output_dir=tmp_path, math_type="latex")
```
Comment (matthiaskoenig) [inline src/sbmlutils/test/report/test_latex.py:12]: You also have to test the other math outputs, i.e. 
sbmlreport.create_report(sbml_path=source, output_dir=tmp_path, math_type="pmathml")
sbmlreport.create_report(sbml_path=source, output_dir=tmp_path, math_type="cmathml")

Comment (matthiaskoenig) [reply src/sbmlutils/test/report/test_latex.py:12]: Also you have to assert that the report was created!

**Review by matthiaskoenig — CHANGES_REQUESTED** [review]
See individual comments.
In addition the tests are not passing on windows due to some encoding issue. You can see the error message looking at the output of the failed continious integration: https://github.com/matthiaskoenig/sbmlutils/pull/200/checks?check_run_id=1898860072
```
 ____________________________ test_report[source0] _____________________________
  
  source = WindowsPath('D:/a/sbmlutils/sbmlutils/src/sbmlutils/test/data/models/repressilator/BIOMD0000000012_urn.xml')
  tmp_path = WindowsPath('C:/Users/runneradmin/AppData/Local/Temp/pytest-of-unknown/pytest-0/test_report_source0_0')
  
      @pytest.mark.parametrize("source", [REPRESSILATOR_SBML])
      def test_report(source, tmp_path):
  >       sbmlreport.create_report(sbml_path=source, output_dir=tmp_path, math_type="latex")
  
  src\sbmlutils\test\report\test_latex.py:12: 
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  src\sbmlutils\report\sbmlreport.py:141: in create_report
      f_html.write(html)
  _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
  
  self = <encodings.cp1252.IncrementalEncoder object at 0x000001DB2E7F05C8>
  input = '<!doctype html>\r\n<html lang="en">\r\n<head>\r\n    <meta charset="utf-8">\r\n    <meta http-equiv="X-UA-Compatible"...></i> sbmlutils</a></p>\r\n            </div>\r\n\r\n        </div>\r\n    </div>\r\n</main>\r\n</body>\r\n</html>\r\n'
  final = False
  
      def encode(self, input, final=False):
  >       return codecs.charmap_encode(input,self.errors,encoding_table)[0]
  E       UnicodeEncodeError: 'charmap' codec can't encode character '\u03c4' in position 23862: character maps to <undefined>
  
  c:\hostedtoolcache\windows\python\3.7.9\x64\lib\encodings\cp1252.py:19: UnicodeEncodeError
```

**Comment by matthiaskoenig** [comment]
You can already see the Latex rendering is a great improvement of the formulas. The windows issues are definitely some encoding issues and you have to make sure the encodings are set correctly.
Most likely on ` f_html.write(html)`.
Google the issue and figure out what the problem is. Windows has a different standard encoding, so the following should be correct:
https://stackoverflow.com/questions/27092833/unicodeencodeerror-charmap-codec-cant-encode-characters

**Comment by matthiaskoenig** [comment]
@sankha555 Also you should watch the project (this is a link on the landing page). Then you are automatically updated about changes not only related to your issues.

**Comment by sankha555** [comment]
@matthiaskoenig Thanks a lot for the detailed  feedback and resources. I have taken a note of all your suggestions and instructions. I would make a combined push to include all the suggested points and also address the encoding issue of Windows by researching on it more. I would try to make these changes by tomorrow evening (I have a university course quiz tomorrow morning, will start working right after that). I would also fix the flake8 issues in this and other modules that you had raised an issue for.

The Latex rendered formulae look really good now, and have improved the readability and look of the formulae. This issue was a really good exercise to work on. Hoping to complete the patch tomorrow and completing the issue. I have also added the repository to my watch list with all updates on and the email updates are working for all changes now. 

Thanks a lot again for the feedback!

**Comment by matthiaskoenig** [comment]
The flake8 issues are fixed in develop. So make sure to rebase against the develop branch. Just send me a quick comment when the pull request is ready for review.

**Comment by sankha555** [comment]
@matthiaskoenig I have incorporated the changes in the PR, please take a look. The windows test error has also been fixed. Please let me know if I should make anymore changes. Thanks!

**Review by matthiaskoenig — CHANGES_REQUESTED** [review]
Please add the other test models to the parametrized tests. These should work with the bugfixes in develop. If all the test models pass the tests I merge this.

**Comment by sankha555** [comment]
@matthiaskoenig I added the other 6 test models to the parameterized tests. All of them pass the new math types too, except for one. The model ```GLUCOSE_SBML``` fails the tests for ```pmathml``` and ```latex``` math types, and generates the following error log :

```

source = PosixPath('/Users/maniklaldas/Desktop/Misc/sbmlutils/src/sbmlutils/test/data/models/glucose/Hepatic_glucose_3.xml')
tmp_path = PosixPath('/private/var/folders/xy/d8k2rr415_j4h5zphvhr8vg00000gn/T/pytest-of-maniklaldas/pytest-24/test_report_latex_source3_0')

    @pytest.mark.parametrize("source", [BASIC_SBML,
                                        DEMO_SBML,
                                        GALACTOSE_SINGLECELL_SBML,
                                        GLUCOSE_SBML,
                                        GZ_SBML,
                                        VDP_SBML,
                                        REPRESSILATOR_SBML])
    def test_report_latex(source, tmp_path):
>       sbmlreport.create_report(sbml_path=source, output_dir=tmp_path, math_type="latex")

test_math_type.py:29: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../report/sbmlreport.py:155: in create_report
    html = _create_html(doc, basename, html_template=template, math_type=math_type)
../../report/sbmlreport.py:250: in _create_html
    "functions": listOfFunctions_dict(model, math_type),
../../report/sbmlreport.py:502: in listOfFunctions_dict
    info["math"] = math(item, math_type)
../../report/sbmlreport.py:969: in math
    latex_str = mathml.astnode_to_latex(item.getMath())
../../report/mathml.py:224: in astnode_to_latex
    expr = astnode_to_expression(astnode)
../../report/mathml.py:67: in astnode_to_expression
    return formula_to_expression(formula)
../../report/mathml.py:88: in formula_to_expression
    expr = sympify(formula)
../../../../../../sb/lib/python3.7/site-packages/sympy/core/sympify.py:480: in sympify
    expr = parse_expr(a, local_dict=locals, transformations=transformations, evaluate=evaluate)
../../../../../../sb/lib/python3.7/site-packages/sympy/parsing/sympy_parser.py:1008: in parse_expr
    return eval_expr(code, local_dict, global_dict)
../../../../../../sb/lib/python3.7/site-packages/sympy/parsing/sympy_parser.py:903: in eval_expr
    code, global_dict, local_dict)  # take local objects in preference
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   TypeError: __new__() missing 1 required positional argument: 'expr'

<string>:1: TypeError
--------------------------------------------------------------------------- Captured stdout call ---------------------------------------------------------------------------

--------------------------------------------------------------------------------
/Users/maniklaldas/Desktop/Misc/sbmlutils/src/sbmlutils/test/data/models/glucose/Hepatic_glucose_3.xml
valid                    : TRUE
check time (s)           : 0.469
--------------------------------------------------------------------------------
```
The  run fails when the ```eval_expr``` method is called from ```sympy```, where it reports missing expression argument. I printed the formula to the console logs and it is the following:

```
lambda(x, y, Piecewise((x, x > y),(y, True)))
```

I think that due to some reason this formula is not being evaluated by sympy in one of the intermediate methods (as given in the error logs), and due to this the tests are failing for this particular model. The cmathml test still passes because it is using the cmathml conversion from ```formating.py```. 

Please let me know if something can be done to fix the issue for this model. I would run the tests again on it once we find a solution and then all the models will have passed the tests. Thanks!

**Comment by matthiaskoenig** [comment]
I will fix the issue and merge the pull request. That is why we have to
test against the biomodels also.

**Comment by sankha555** [comment]
I understand the point, doing the tests on the biomodels helped to identify the potential bug fixes that could be made. I am also working on the curated biomodels integration into the tests, hopefully that will bring out more points that could be worked upon and hope to get most of the tests passed with all those models.

**Comment by sankha555** [comment]
@matthiaskoenig There is one change that I would make regarding a redundant check in the tests. I would remove the XML file check (present right after the HTML report check), as it was basically just checking the existence of the original source file. I would just let the HTML report generation assertion to stay.

I would make the change in a commit now. In case you think I should make any change please let me know and I would do it. After that, if you think everything is fine, we can squash the commits and you could then merge the PR. Thanks!

**Review by matthiaskoenig — APPROVED** [review]
I cleaned up the code of yours.
You had a lot of code duplication in the testing and also in the math.
Reduce code duplication whenever possible, e.g., by writing a `check_function` in the tests.

**Comment by sankha555** [comment]
@matthiaskoenig Thanks a lot for the help and guidance. I have taken a note of the changes you have made and have got an idea of how and where to reduce code duplication. I would make sure to follow this guideline in my future pull requests. Thanks!

**Pull request merged**